File: wrappers/dotnet/PrinterDriver.Tests/DrawerTests.cs
namespace PrinterDriver.Tests;

/// <summary>
/// M14 — the cash drawer, through the managed wrapper (docs/cash-drawer.md).
/// </summary>
/// <remarks>
/// The wrapper holds no drawer logic: the sequence, the refusals and the polarity all live
/// in the core. What is under test here is that none of it is lost on the way across —
/// that a refusal is still a refusal, that <c>KickSentUnverified</c> does not quietly
/// become a success, and that an uncalibrated switch still reports a level rather than a
/// direction.
/// </remarks>
public sealed class DrawerTests
{
    [Fact]
    public void AVerifiedOpenIsReportedOnlyWhenTheSwitchMoves()
    {
        using var driver = TestDriver.Open();
        var till = driver.AddScripted("drawer", "till");

        var caps = till.DrawerCapabilities;
        Assert.True(caps.Present);
        Assert.True(caps.Kickable);
        Assert.Equal(DrawerPortStandard.Epson24V6P6C, caps.PortStandard);
        Assert.Equal(DrawerKickMethod.EpsonEscP, caps.KickMethod);
        Assert.Equal(DrawerStatusMethod.GsR2, caps.StatusMethod);
        Assert.Equal(24, caps.Voltage);
        Assert.Equal(1000, caps.MaxCurrentMa);
        Assert.Equal(2, caps.ChannelCount);
        // The single fact that separates this arrangement from Star's.
        Assert.Equal(3, caps.SensorPin);
        Assert.Equal(Provenance.Documented, caps.ElectricalProvenance);
        Assert.Equal(Provenance.Documented, caps.CommandsProvenance);

        var result = till.OpenDrawer();
        Assert.Equal(DrawerState.Closed, result.PreviousState);
        Assert.Equal(DrawerState.OpenVerified, result.State);
        Assert.True(result.Verified);
        Assert.Equal(1, result.Channel);
        Assert.Equal(200, result.PulseMs);
        Assert.Equal((nuint)1, TestNative.pd_test_drawer_kicks(till.Handle));
        Assert.Equal(1, TestNative.pd_test_drawer_is_open(till.Handle));

        // Step 1 of the sequence: a drawer that is already out is never pulsed again.
        var again = till.OpenDrawer(channel: 1, pulseMs: 120);
        Assert.Equal(DrawerState.Open, again.State);
        Assert.Equal(0, again.PulseMs);
        Assert.False(again.Verified);
        Assert.Equal((nuint)1, TestNative.pd_test_drawer_kicks(till.Handle));
    }

    [Fact]
    public void ALockedDrawerIsFailedToOpenAndNotASuccess()
    {
        using var driver = TestDriver.Open();
        var till = driver.AddScripted("drawer-locked", "locked");

        var result = till.OpenDrawer(channel: 2, pulseMs: 120);
        Assert.Equal(DrawerState.Closed, result.PreviousState);
        Assert.Equal(DrawerState.FailedToOpen, result.State);
        Assert.False(result.Verified);
        Assert.Equal(2, result.Channel);
        Assert.Equal(120, result.PulseMs);
        // The pulse was real; the drawer was not.
        Assert.Equal((nuint)1, TestNative.pd_test_drawer_kicks(till.Handle));
        Assert.Equal(0, TestNative.pd_test_drawer_is_open(till.Handle));
    }

    [Fact]
    public void AnUnclassifiedPortIsRefusedWithoutWritingAByte()
    {
        using var driver = TestDriver.Open();
        var till = driver.AddScripted("drawer-unknown-port", "unclassified");

        var caps = till.DrawerCapabilities;
        Assert.Equal(DrawerPortStandard.Unknown, caps.PortStandard);
        Assert.False(caps.Kickable);

        var result = till.OpenDrawer();
        Assert.Equal(DrawerState.Unknown, result.State);
        Assert.Equal(0, result.PulseMs);
        // RJ11/RJ12-looking drawer connectors are not a universal electrical standard, and
        // an unclassified port is never energised.
        Assert.Equal((nuint)0, TestNative.pd_test_drawer_kicks(till.Handle));

        // Reading the switch is still safe on the same hardware: it asks a question and
        // energises nothing, which is what makes it the probe's non-destructive half.
        var reading = till.ReadDrawerSensor(500);
        Assert.True(reading.Available);
        Assert.True(reading.Answered);
        Assert.Equal((nuint)0, TestNative.pd_test_drawer_kicks(till.Handle));
    }

    [Fact]
    public void AnUncalibratedSwitchReportsALevelAndNotADirection()
    {
        using var driver = TestDriver.Open();
        var till = driver.AddScripted("drawer-uncalibrated", "uncalibrated");
        Assert.False(till.IsDrawerPolarityCalibrated);

        // The operator procedure, without the prompts.
        TestNative.pd_test_set_drawer_open(till.Handle, 0);
        var shut = till.ReadDrawerSensor(500);
        Assert.True(shut.Answered);
        Assert.True(shut.NeedsCalibration);
        Assert.Equal(false, shut.PinHigh);
        // Whether the line reads high or low when the drawer is open depends on the drawer
        // that is plugged in, so until it is measured there is no interpretation.
        Assert.Equal(DrawerState.Unknown, shut.State);

        TestNative.pd_test_set_drawer_open(till.Handle, 1);
        var open = till.ReadDrawerSensor(500);
        Assert.Equal(true, open.PinHigh);
        Assert.NotEqual(shut.PinHigh, open.PinHigh);
        Assert.Equal((nuint)0, TestNative.pd_test_drawer_kicks(till.Handle));

        // In-memory driver: the calibration applies to this process, and says so by
        // returning false rather than pretending it was persisted.
        Assert.False(till.CalibrateDrawerPolarity(open.PinHigh ?? true));
        Assert.True(till.IsDrawerPolarityCalibrated);
        Assert.True(till.DrawerHighMeansOpen);

        var after = till.ReadDrawerSensor(500);
        Assert.False(after.NeedsCalibration);
        Assert.Equal(DrawerState.Open, after.State);
    }

    [Fact]
    public void ALabelPrinterHasNoDrawerAndFiresNothing()
    {
        using var driver = TestDriver.Open();
        // Zebra speaks ZPL and CPCL, has no drawer port, and every job on it is refused
        // before a byte is written. The drawer call is the same refusal.
        var labels = driver.AddPrinterTcp(new TcpPrinterConfig(
            "192.0.2.60", PrinterId: "labels", ProfileId: "zebra_zq600_plus"));

        var caps = labels.DrawerCapabilities;
        Assert.False(caps.Present);
        Assert.Equal(DrawerKickMethod.Unsupported, caps.KickMethod);
        Assert.False(caps.Kickable);

        var result = labels.OpenDrawer();
        Assert.Equal(DrawerState.Unknown, result.State);
        Assert.Equal(0, result.PulseMs);
    }
}

File: wrappers/swift/Tests/PrinterDriverTests/DrawerTests.swift
import CPrinterDriver
import CPrinterDriverTestSupport
import XCTest

@testable import PrinterDriver

/// M14 — the cash drawer, through the Swift wrapper (docs/cash-drawer.md).
///
/// The wrapper contains no drawer logic — the sequence, the refusals and the polarity all
/// live in the core — so what is under test here is that none of that is *lost* on the way
/// across: that a refusal is still a refusal, that `kickSentUnverified` does not quietly
/// become a success, and that an uncalibrated switch still reports a level rather than a
/// direction.
final class DrawerTests: XCTestCase {

  func testAVerifiedOpenIsReportedOnlyWhenTheSwitchMoves() throws {
    let driver = try PrinterDriver(fsyncDisabled: true)
    let till = try driver.scriptedPrinter(id: "till", .drawer)

    let caps = till.drawerCapabilities
    XCTAssertTrue(caps.isPresent)
    XCTAssertTrue(caps.isKickable)
    XCTAssertEqual(caps.portStandard, .epson24V6P6C)
    XCTAssertEqual(caps.kickMethod, .epsonEscP)
    XCTAssertEqual(caps.statusMethod, .gsR2)
    XCTAssertEqual(caps.voltage, 24)
    XCTAssertEqual(caps.maxCurrentMilliamps, 1000)
    XCTAssertEqual(caps.channelCount, 2)
    // The single fact that separates this arrangement from Star's.
    XCTAssertEqual(caps.sensorPin, 3)
    XCTAssertEqual(caps.electricalProvenance, .documented)
    XCTAssertEqual(caps.commandsProvenance, .documented)

    let result = till.openDrawer()
    XCTAssertEqual(result.previousState, .closed)
    XCTAssertEqual(result.state, .openVerified)
    XCTAssertTrue(result.isVerified)
    XCTAssertEqual(result.channel, 1)
    XCTAssertEqual(result.pulseMilliseconds, 200)
    XCTAssertEqual(till.scriptedDrawerKicks, 1)
    XCTAssertTrue(till.scriptedDrawerIsOpen)
    XCTAssertEqual(result.state.abiName, "OpenVerified")

    // Step 1 of the sequence: a drawer that is already out is never pulsed again.
    let again = till.openDrawer(channel: 1, pulseMilliseconds: 120)
    XCTAssertEqual(again.state, .open)
    XCTAssertEqual(again.pulseMilliseconds, 0)
    XCTAssertFalse(again.isVerified)
    XCTAssertEqual(till.scriptedDrawerKicks, 1)
  }

  func testALockedDrawerIsFailedToOpenAndNotASuccess() throws {
    let driver = try PrinterDriver(fsyncDisabled: true)
    let till = try driver.scriptedPrinter(id: "locked", .drawerLocked)

    let result = till.openDrawer(channel: 2, pulseMilliseconds: 120)
    XCTAssertEqual(result.previousState, .closed)
    XCTAssertEqual(result.state, .failedToOpen)
    XCTAssertFalse(result.isVerified)
    XCTAssertEqual(result.channel, 2)
    XCTAssertEqual(result.pulseMilliseconds, 120)
    // The pulse was real; the drawer was not.
    XCTAssertEqual(till.scriptedDrawerKicks, 1)
    XCTAssertFalse(till.scriptedDrawerIsOpen)
  }

  func testAnUnclassifiedPortIsRefusedWithoutWritingAByte() throws {
    let driver = try PrinterDriver(fsyncDisabled: true)
    let till = try driver.scriptedPrinter(id: "unclassified", .drawerUnknownPort)

    let caps = till.drawerCapabilities
    XCTAssertEqual(caps.portStandard, .unknown)
    XCTAssertFalse(caps.isKickable)

    let result = till.openDrawer()
    XCTAssertEqual(result.state, .unknown)
    XCTAssertEqual(result.pulseMilliseconds, 0)
    // RJ11/RJ12-looking drawer connectors are not a universal electrical standard, and
    // an unclassified port is never energised.
    XCTAssertEqual(till.scriptedDrawerKicks, 0)

    // Reading the switch is still safe on the same hardware: it asks a question and
    // energises nothing, which is what makes it the probe's non-destructive half.
    let reading = till.readDrawerSensor(timeoutMilliseconds: 500)
    XCTAssertTrue(reading.isAvailable)
    XCTAssertTrue(reading.didAnswer)
    XCTAssertEqual(till.scriptedDrawerKicks, 0)
  }

  func testAnUncalibratedSwitchReportsALevelAndNotADirection() throws {
    let driver = try PrinterDriver(fsyncDisabled: true)
    let till = try driver.scriptedPrinter(id: "uncalibrated", .drawerUncalibrated)
    XCTAssertFalse(till.isDrawerPolarityCalibrated)

    // The operator procedure, without the prompts.
    till.scriptedSetDrawerOpen(false)
    let shut = till.readDrawerSensor(timeoutMilliseconds: 500)
    XCTAssertTrue(shut.didAnswer)
    XCTAssertTrue(shut.needsCalibration)
    XCTAssertEqual(shut.pinHigh, false)
    // Star documents that the meaning of the level depends on the attached drawer, so
    // until it is measured there is no interpretation to report.
    XCTAssertEqual(shut.state, .unknown)

    till.scriptedSetDrawerOpen(true)
    let open = till.readDrawerSensor(timeoutMilliseconds: 500)
    XCTAssertEqual(open.pinHigh, true)
    XCTAssertNotEqual(shut.pinHigh, open.pinHigh, "a switch that moved")
    XCTAssertEqual(till.scriptedDrawerKicks, 0, "reading never fires anything")

    // In-memory driver: the calibration applies to this process and says so by returning
    // false rather than pretending it was persisted.
    XCTAssertFalse(till.calibrateDrawerPolarity(highMeansOpen: open.pinHigh ?? true))
    XCTAssertTrue(till.isDrawerPolarityCalibrated)
    XCTAssertTrue(till.drawerHighMeansOpen)

    let after = till.readDrawerSensor(timeoutMilliseconds: 500)
    XCTAssertFalse(after.needsCalibration)
    XCTAssertEqual(after.state, .open)
  }

  func testALabelPrinterHasNoDrawerAndFiresNothing() throws {
    let driver = try PrinterDriver(fsyncDisabled: true)
    // Zebra speaks ZPL and CPCL, has no drawer port, and every job on it is refused
    // before a byte is written. The drawer call is the same refusal.
    let labels = try driver.printer(.tcp(host: "192.0.2.60"), profile: "zebra_zq600_plus")

    let caps = labels.drawerCapabilities
    XCTAssertFalse(caps.isPresent)
    XCTAssertEqual(caps.kickMethod, .unsupported)
    XCTAssertFalse(caps.isKickable)

    let result = labels.openDrawer()
    XCTAssertEqual(result.state, .unknown)
    XCTAssertEqual(result.pulseMilliseconds, 0)
  }
}
